i6_models/parts/factored_hybrid/diphone.py

```diff
@@ -0,0 +1,167 @@
+__all__ = [
+    "DiphoneLogitsV1",
+    "DiphoneProbsV1",
+    "DiphoneBackendV1Config",
+    "DiphoneBackendV1",
+]
+
+from dataclasses import dataclass
+from typing import Callable, Optional, Union
+
+import torch
+from torch import nn, Tensor
+import torch.nn.functional as F
+
+from i6_models.config import ModelConfiguration
+
+from .util import PhonemeStateClassV1, get_center_dim, get_mlp
+
+
+@dataclass
+class DiphoneLogitsV1:
+    """outputs of a diphone factored hybrid model"""
+
+    embeddings_left_context: Tensor
+    """the embedded left context values"""
+
+    output_center: Tensor
+    """center output"""
+
+    output_left: Tensor
+    """left output"""
+
+
+class DiphoneProbsV1(DiphoneLogitsV1):
+    """marker class indicating that the output tensors store probabilities and not logits"""
+
+
+@dataclass
+class DiphoneBackendV1Config(ModelConfiguration):
+    """
+    Attributes:
+        context_mix_mlp_dim: inner dimension of the context mixing MLP layers
+        context_mix_mlp_num_layers: how many hidden layers on the MLPs there should be
+        dropout: dropout probabilty
+        left_context_embedding_dim: embedding dimension of the left context
+            values. Good choice is in the order of n_contexts.
+        n_contexts: the number of raw phonemes/acoustic contexts
+        num_hmm_states_per_phone: the number of HMM states per phoneme
+        num_inputs: input dimension of the backend, must match w/ output dimension
+            of main encoder (e.g. Conformer)
+        phoneme_state_class: the phoneme state augmentation to apply
+        activation: activation function to use in the context mixing MLP.
+    """
+
+    left_context_embedding_dim: int
+    n_contexts: int
+    num_hmm_states_per_phone: int
+    phoneme_state_class: Union[int, PhonemeStateClassV1]
+
+    activation: Callable[[], nn.Module]
+    context_mix_mlp_dim: int
+    context_mix_mlp_num_layers: int
+    dropout: float
+    num_inputs: int
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+
+        assert self.context_mix_mlp_dim > 0
+        assert self.context_mix_mlp_num_layers > 0
+        assert self.n_contexts > 0
+        assert self.num_hmm_states_per_phone > 0
+        assert self.num_inputs > 0
+        assert self.left_context_embedding_dim > 0
+        assert 0.0 <= self.dropout <= 1.0, "dropout must be a probability"
+
+
+class DiphoneBackendV1(nn.Module):
```

Comment:
I am a bit unhappy about the name "backend" as that term says exactly nothing. I once talked to guys from Paderborn who were doing denoising and source separation of audio spectra and when they said "backend" they meant "feature extraction and other ASR model stuff" 🤷 

I was trying to find a good name that implies that here embeddings are converted to (log-)probabilities.. but I failed.. pls help :)

Reply:
+1 for Willis statement. Please really do not call it Backend. Maybe `DiphoneOutputBlock`?

Reply:
+1 for Moritz idea.
In defense of the naming, in the literature and commonly within ASR community, the feature extraction and all respective methods such as source separation, beam forming, and enhancement are denoted as ASR frontend and the model itself is called backend. I find the naming well chosen since it reminds the user that this can be combined with any encoder architecture or speech representation h(x).
The new generation calls the similar part of an RNN-T as the "decoder", which is certainly not a better term.

Reply:
We just had a live discussion with @JackTemaki and @ZijYang and we think a nice name for this can be `DiphoneJoinerBlock` reminding the similar RNN-T case

Reply:
I like "Joiner" better than "Backend" and "Output" most of them. Although this is not necessarily the (final) output of the model.

`DiphoneProbabilityBlock` ? or even just (`Factored`-)`DiphoneBlock` ?

Reply:
I also like `FactoredDiphoneBlock`. 
BTW, one feedback I got from people outside our research group is that there is a marketing problem with the name of factored hybrid, and if we had a fancy new name people would be more open to try it out. Independently of the name, using up-to-date terminologies that do not remind the GMM-HMM systems is preferable, IMO.

Reply:
> if we had a fancy new name people would be more open to try it out. 

_seigh_

> Independently of the name, using up-to-date terminologies that do not remind the GMM-HMM systems is preferable, IMO.

Nothing inherently needs this to be for "phone"s it could also be FactoredDichar or DiBPE.. (except for the adherence to rasrs phoneme numbering scheme in `PhonemeStateClassV1`).. 

So should we drop the "Phone" branding?^^